File: blog/src/lib/models/tweets.ts
/** Represents a completed image or video attached to a tweet. */
export interface TweetMedia {
	id: string;
	publicUrl: string;
	contentType: string;
	mediaKind: 'image' | 'video';
	position: number;
}

/** Represents a public short-form post. */
export interface Tweet {
	id: string;
	authorId: string;
	body: string;
	tags: string[];
	createdAt: string;
	media: TweetMedia[];
}

/** Represents one cursor-paginated timeline response. */
export interface TweetPage {
	tweets: Tweet[];
	nextBefore: string | null;
	hasMore: boolean;
}

/** Contains the direct-upload details for a pending attachment. */
export interface TweetUpload {
	mediaId: string;
	uploadUrl: string;
	publicUrl: string;
}

File: backend/src/models/tweets.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// A media attachment displayed with a tweet.
#[derive(Debug, Serialize, sqlx::FromRow)]
#[serde(rename_all = "camelCase")]
pub struct TweetMedia {
    pub id: Uuid,
    pub public_url: String,
    pub content_type: String,
    pub media_kind: String,
    pub position: i32,
}

/// A public short-form post and its completed attachments.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Tweet {
    pub id: Uuid,
    pub author_id: Uuid,
    pub body: String,
    pub tags: Vec<String>,
    pub created_at: DateTime<Utc>,
    pub media: Vec<TweetMedia>,
}

/// A cursor-paginated tweet timeline.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TweetPage {
    pub tweets: Vec<Tweet>,
    pub next_before: Option<Uuid>,
    pub has_more: bool,
}

/// Supplies a completed media batch for a newly published tweet.
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateTweetRequest {
    pub body: String,
    pub tags: Vec<String>,
    pub media_ids: Vec<Uuid>,
}
